File: 3/3000s/3355/ZeroArrayTransformationI.cpp
/*
https://leetcode.com/problems/zero-array-transformation-i/description/?envType=daily-question&envId=2025-05-20
*/

class Solution {
public:
    bool isZeroArray(vector<int>& nums, vector<vector<int>>& queries) {
        int sz1=nums.size(), sz2=queries.size();
        vector<int>diffVec(sz1+1, 0);
        for(int i=0; i<sz2; ++i){
            int lft=queries[i][0], rght=queries[i][1];
            diffVec[lft]++, diffVec[rght+1]--;
        }
        vector<int>allAvlOperations;
        int curOperations=0;
        for(auto &vals: diffVec){
            curOperations+=vals;
            allAvlOperations.push_back(curOperations);
        }
        for(int i=0; i<sz1; ++i){
            if(allAvlOperations[i]<nums[i]) return false;
        }
        return true;
    }
};

File: 3/3000s/3355/ZeroArrayTransformationI.java
/*
https://leetcode.com/problems/zero-array-transformation-i/description/?envType=daily-question&envId=2025-05-20
*/

class Solution {
    public boolean isZeroArray(int[] nums, int[][] queries) {
        int lengthOfNums=nums.length, lengthOfQueries=queries.length;
        ArrayList<Integer>differenceVector=new ArrayList<>();
        for(int i=0; i<=lengthOfNums; ++i){
            differenceVector.add(0);
        }
        for(int i=0; i<lengthOfQueries; ++i){
            int left=queries[i][0], right=queries[i][1];
            differenceVector.set(left, differenceVector.get(left)+1);
            differenceVector.set(right+1, differenceVector.get(right+1)-1);
        }
        ArrayList<Integer>allAvailableOperations=new ArrayList<>();
        int currentOperation=0;
        for(int diff: differenceVector){
            currentOperation+=diff;
            allAvailableOperations.add(currentOperation);
        }
        for(int i=0; i<lengthOfNums; ++i){
            if(allAvailableOperations.get(i)<nums[i]) return false;
        }
        return true;
    }
}